### PR TITLE
chore: Add -slim Dockerfile and update Docker workflows

### LINKED ISSUE #1007: Reimplement Docker Hub Github Actions

### LINKED ISSUE #1010: Docker images don't publish on our release tags
Since we've added Release Please, a workflow is creating our tags, which can't trigger another workflow to run (to avoid the chance that it causes a circular trigger). However, we can trigger it with workflow_dispatch to create a tagged release.

### PR DESCRIPTION
**Pull request by ospencer**
Closes #1007
Closes #1010 

This PR updates the `docker/build-push-action` GitHub action to v2 (which allows us to publish to DockerHub again) and adds a new `Dockerfile-slim` which provides a minimal environment that can compile Grain programs. The slim version is about 10x smaller!

image

You can see a [successful run of the GitHub workflow](https://github.com/grain-lang/grain/runs/4243823185?check_suite_focus=true), [images published on DockerHub](https://hub.docker.com/repository/docker/grainlang/grain/tags?page=1&ordering=last_updated&name=oscar-docker), and [on the GitHub Container Registry](https://github.com/grain-lang/grain/pkgs/container/grain). Note that we previously were (accidentally) publishing on GHCR as `grain-lang/grain/grain` rather than `grain-lang/grain`.

In addition to the push on `main`, there's also a `workflow_dispatch` event that gets triggered on release. Shoutout to @phated for making it easy for me to figure out how to do that 😤 

You can see that the tag is handled properly via [this workflow](https://github.com/ospencer/grain/runs/4244964750?check_suite_focus=true) (and that it's ignored on `main` via [this workflow](https://github.com/ospencer/grain/runs/4244950280?check_suite_focus=true)), but I've also got some screenshots here to save you some clicks:

For the standard image:
image

For the slim image:
image

When just running on a push to `main`:
image

The docker metadata action is quite nice, as you can see that it will also handle having a release tag (i.e. `grain:0.4.4`) and also a tag for the minor for automatic patch updates (i.e. `grain:0.4`).

Besides that, the main Dockerfile was simplified quite a bit. The slim version is similar, but it has to manually install some packages and then does a bunch of cleanup.

### CODE REVIEW

**Review by marcusroberts — APPROVED** [review]
That is sweet!

File: .github/workflows/docker-publish.yml [hunk 1]
```diff
@@ -1,30 +1,105 @@
...
+          registry: ghcr.io
+          username: ${{ github.actor }}
+          password: ${{ secrets.GITHUB_TOKEN }}
+
+      - name: Build and push image
+        uses: docker/build-push-action@v2
+        with:
+          context: .
+          file: Dockerfile
+          push: ${{ github.event_name != 'pull_request' }}
```
Comment (phated) [inline .github/workflows/docker-publish.yml:54]: If we only run this on `main`, why do we need to filter on `!= pull_request`?

File: .dockerignore [hunk 2]
```diff
@@ -21,6 +21,7 @@ script/node_modules
 **/*.log
 _build/
 */_build/
+cli/bin/*.exe
```
Comment (phated) [inline .dockerignore:24]: Should also exclude the bc.js stuff
```suggestion
cli/bin/*.exe
cli/bin/*.bc.js
```

File: .github/workflows/docker-publish.yml [hunk 3]
```diff
@@ -1,30 +1,105 @@
...
-          registry: docker.pkg.github.com
-          repository: grain-lang/grain/grain
-          tag_with_ref: true
+
+      - name: Build and push slim image
+        uses: docker/build-push-action@v2.7.0
+        with:
+          context: .
+          file: Dockerfile-slim
+          push: ${{ github.event_name != 'pull_request' }}
```
Comment (phated) [inline .github/workflows/docker-publish.yml:103]: same question

**Review by phated — COMMENTED** [review]
This looks really solid to me! There are a couple things I've noted, and I believe you should add "Closes X" for possibly 2-3 issues in the tracker (?)

File: .github/workflows/docker-publish.yml (hunk 1 above)
Comment (ospencer) [reply .github/workflows/docker-publish.yml:54]: copypasta 🙈

**Comment by ospencer** [comment]
@phated Should be all set here, and added some issues to close!

**Review by phated — APPROVED** [review]
:shipit: Excellent work!

**Pull request merged**